README
# slurm - yet another network load monitor

slurm started as a FreeBSD port of the Linux ppp link monitor called pppstatus
by Gabriel Montenegro. Hendrik Scholz ripped of the ppp dependent parts and the
email checks to turn in into a generic network load monitor for *BSD, Linux,
HP-UX and Solaris.

## features:
 - "realtime" traffic statistics
 - three graph modes: combined RX and TX and two split views
 - can monitor any network device
 - curses ascii graphics
 - ascii theme support

slurm was tested on:
 - Debian GNU/Linux 8.0
 - FreeBSD 10
 - CentOS 6/7

## Compiling:
 - Install cmake and libncurses5-dev e.g.:
on Debian (if you don't want to use the Debian package of slurm)

         apt-get install build-essential
         apt-get install cmake libncurses5-dev
or on CentOS / RedHat

        yum groupinstall 'Development Tools'
        yum install cmake ncurses-devel

 - Create a build directory

        mkdir _build
 - Run cmake to create the Makefiles

        cd _build/
        cmake ..
 - Run make to build the binary

        make
 - Install the program

        sudo make install
 - Try it!

        slurm -i eth0 # or whatever your interface is

Please report bugs and feature requests in the github bugtracker:
 website: https://github.com/mliska/slurm

License:
 As pppstatus is licensed under the GPL, slurm is too.


ChangeLog
2019-01-14 Milos Liska <mliska@gmail.com>
	* tag version 0.5.0
	* Added iface speed detection on Linux
	* Use bits per second instead of bytes per second
	* Better usage of the 80x25 default terminal real estate
	* Added speed and time axes labels to the graph
	* Fixed graph width rendering (no more artefacts in the last column)
	* Use different default SYMBOL_TRAFFIC and SYMBOL_NONTRAFFIC (based on bmon)
	* Fixed colors rendering if SYMBOL_NONTRAFFIC is not set to space
	* Updated manpage

2015-08-16 Matthias Schmitz <matthias@sigxcpu.org>
	* tag version 0.4.3
	* switch to cmake as build system

2014-10-14 Matthias Schmitz <matthias@sigxcpu.org>
        * tag version 0.4.2
	* Always use "defined" when testing for a preprocessor macro

2011-09-05 Matthias Schmitz <matthias@sigxcpu.org>
	* tag version 0.4.0
	* Remove gnu build system, use scons instead
	* Add support for GNU/kFreeBSD
	* Probably break support for any other os than Linux and kFreeBSD (
	  will fixed soon )


- tag SLURM-0_3_3

2004-10-12  Hendrik Scholz  <hscholz@raisdorf.net>

	* Makefile:
	* config.guess:
	* config.sub:

	- update config.* to allow configure run on DragonFly BSD (i386 only)


2004-09-28  Hendrik Scholz  <hscholz@raisdorf.net>

	* Makefile:
	* configure:
	* configure.in:

	bump to VERSION 0.3.3

2004-09-28  Hendrik Scholz  <hscholz@raisdorf.net>

	* README: minor OS updates
	* os.h: allow compilation on Sun Forte CC systems again
	* slurm.c: (main):
	* theme.c: (theme_readfile): remove NOTICE error
	* themes/black.theme: new theme

2004-09-27  Hendrik Scholz <hscholz@raisdorf.net>

	* src/if_media.c: (get_if_speed):

	- use better CPP magic to detect OS features
	- allow compilation on a few OpenBSD systems again

2004-05-27  Hendrik Scholz  <hscholz@raisdorf.net>

	* .cvsignore:
	* os.h:
	* src/if_media.c: (get_if_speed):

	- add validinterface() check on Mac OS X

2004-02-13  Hendrik Scholz  <hscholz@raisdorf.net>

	* Makefile:
	* Makefile.in:
	* aclocal.m4:
	* config.h.in:
	* configure:
	* configure.in:
	* slurm.c: (usage):
	* slurm.h:

	- remove version name from OS for 'darwin' aka Mac OS X
	- s/VERSION/PACKAGE_VERSION/

2004-02-13  Hendrik Scholz  <hscholz@wormulon.net>

	* AUTHORS:
	* COPYING:
	* Changelog:
	* INSTALL:
	* Makefile:
	* Makefile.am:
	* Makefile.in:
	* THEMES:
	* THEMES.txt:
	* aclocal.m4:
	* config.guess:
	* config.h.in:
	* config.sub:
	* configure:
	* configure.in:
	* depcomp:
	* install-sh:
	* missing:
	* mkinstalldirs:

	- major config subsystem replacement
	- mv THEMES THEMES.txt (welcome to Mac OS X :))

2004-02-05  Hendrik Scholz  <hscholz@wormulon.net>

	* src/if_media.c: (get_if_speed):

    - add support for pre 4.5-RELEASE FreeBSD (tested on 4.2-REL :))

2004-02-05  Hendrik Scholz  <hscholz@wormulon.net>

	* FAQ: added Linux media Q/A
	* README: s/raisdorf/wormulon/
	* slurm.1: s/raisdorf/wormulon/, added -L option
	* slurm.c: (slurm_shutdown): s/raisdorf/wormulon/
	* slurm.spec: s/raisdorf/wormulon/
	* theme.c: (theme_readfile): added -D__Debian__ to search for
		themes in /usr/share rather than /usr/local/share

2004-02-05  Hendrik Scholz  <hscholz@raisdorf.net>

	* Makefile:
	* Makefile.in: add -DWITH_LED per default
	* slurm.c: (update_stat_large), (update_stat_split),
	(update_stat_combined), (usage), (draw_face), (main):
	* slurm.h:

	- add -L switch to enable LED

2003-10-07  Hendrik Scholz  <hscholz@raisdorf.net>

	* Makefile:
	* Makefile.in:
	* theme.c: (theme_readfile):

	- reduce overhead in upcoming NetBSD port upgrade

2003-09-18  Hendrik Scholz  <hscholz@raisdorf.net>

	* TODO: upd
	* slurm.spec: initial specfile based on 0.2.3

Thu Sep 11 14:08:07 CEST 2003

    - clean up Solaris code
    - 0.3.1 RELEASE

Wed Sep 10 21:18:32 CEST 2003

    - print usage information if no interface was given at startup

Wed Sep 10 21:03:59 CEST 2003

    - add Mac OS X support (thanks Jan-Eric Kolbe for providing access!)
    - 0.3.1-pre2

Mon Sep  8 23:22:00 CEST 2003

    - calculate traffic right when using -d option

Sat Sep  6 06:01:27 EDT 2003

    - disable LED (set -DWITH_LED to enable them)
    - moved left part of statistics left to allow traffic > 10.000kBytes/s
    - make slurm Gbit-safe (tests with >100MByte/s will be done soon :))
    - make -Wall happy on Linux

Fri Aug 22 06:20:36 EDT 2003

    - fix 32bit signed vs. unsigned bug
    - improve Linux /proc/net/dev parsing

Sat Aug 16 16:06:32 CEST 2003

    - 0.3.0 release
    - again modify sleep stuff to recognize input while sleeping

Wed Aug 13 00:16:57 CEST 2003

    - unify sleep method

Wed Aug 13 00:03:21 CEST 2003

    - add 'make prof' target for profiling and update .cvsignore

Thu Aug  7 04:21:01 CEST 2003

    - use sleep() instead of usleep() for sleep durations longer
      than one second on Linux to prevent wasting cpu cycles
    - update FAQ, README and add THEMES file

Wed Jul  9 18:44:24 CEST 2003

	- add FAQ
	- add -Wall to compile flags

Wed Jul  9 18:39:26 CEST 2003

	- add signalhandler for various signal so killing slurm doesn't
	  leave a scrambled terminal behind
	- add -t option to usage()
	  
Sun May 25 16:44:49 CEST 2003

	- remove -H option and make it default
	- add initial theme engine (only tested on FreeBSD! until now)

Sun May  4 13:55:46 CEST 2003

	- catch up on FreeBSDs include/net/if_media.h since the typos
      have been fixed.

Sun Apr 27 18:12:49 CEST 2003

	- change data_type structure so we won't get into trouble displaying
	  up to 999999.999GB 

Sat Apr 19 20:09:29 CEST 2003

	- add THE missing 'E'

Tue Apr  8 22:58:45 CEST 2003

	- Solaris interface speed detection code added to src/if_media.c

Wed Apr  2 00:00:34 CEST 2003

    - add Solaris support to auto* stuff so slurm can be build on Solaris
      with just ./configure && make

Wed Mar 26 23:43:54 CET 2003

	- add -z command line option and 'z' key to virtually reset all counters

Thu Mar 20 23:23:46 CET 2003

	- fix padding problems with speeds changing dimensions

Mon Feb 24 19:55:56 CET 2003


slurm.h
/******************************************************************************
 *
 * slurm config file, heavily based on the pppstatus version
 * see README file or slurm.c for more information
 *
 * $Id: slurm.h,v 1.42 2004/02/13 05:41:30 hscholz Exp $
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 ***************************************************************************/

#define LED_RX    1
#define LED_TX    2

#ifndef TRUE
#define TRUE 1
#endif
#ifndef FALSE
#define FALSE 0
#endif

#define SYMBOL_TRAFFIC		"|"
#define SYMBOL_NOTRAFFIC	"."

#ifdef __linux__
#define PATH_NET_DEV "/proc/net/dev"
#endif

#define MODE_COMBINED 1
#define MODE_SPLIT 2
#define MODE_LARGE 3

/* slap HPUX with a large trout
 * HPUX 10.xx cannot handle 2^32 int constants, so we have to "tweak" it
 */

#ifdef __HPUX__
#define SNMPMAXBYTES (4294967295 + 1)
#else
#define SNMPMAXBYTES (4294967296ULL)
#endif

#undef MAXHOSTNAMELEN
#define MAXHOSTNAMELEN 64

#define DRAWLEN 16

/* 10th of a second */
#define REFRESH_DEFAULT 1
#define REFRESH_MIN 1
#define REFRESH_MAX 300

#ifndef BUFSIZ
#define BUFSIZ 1024
#warning "setting buf size to 1024"
#endif

typedef struct statvline {
   int hline[76]; /* This will be the horzontal lines */
} statvline[10]; /* ... and this will be vertical lines */
statvline graph;

/* This structure stays the INFO variables */
typedef struct DataStats {
#if (defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L))
    unsigned long long rx_packets;
    unsigned long long rx_errors;
#else
    unsigned long rx_packets;
    unsigned long rx_errors;
#endif
    int rx_over;
#if (defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L))
    unsigned long long tx_packets;
    unsigned long long tx_errors;
#else
    unsigned long tx_packets;
    unsigned long tx_errors;
#endif
    int tx_over;
    double rx_bytes;
    double tx_bytes;
    double rx_bytes_comp;
    double tx_bytes_comp;
    double rx_packets_led;
    double tx_packets_led;
    unsigned long connect_time;
    unsigned long current_time;
    float top_speed;
    int online_days;
    int online_hour;
    int online_min;
    int online_sec;
	unsigned long rx_packets_off;
	unsigned long rx_errors_off;
    int rx_over_off;
    unsigned long tx_packets_off;
    unsigned long tx_errors_off;
    int tx_over_off;
    double rx_bytes_off;
    double tx_bytes_off;
    double rx_bytes_comp_off;
    double tx_bytes_comp_off;
} DataStats;
DataStats stats;

typedef struct IfData{
    char if_name[IFNAMSIZ];       /* The device name given as start parameter*/
    int  if_speed;          /* The Interface speed */
    char if_speedstring[12]; /* the measuring unit like Mbit, kbit */
    int  if_id;             /* The ID which the interface inside the OS has */
    int  if_amount;         /* The amount of all interfaces available */
    int  if_valid;          /* 1 = selected interface exists
                             * 0 = interfaces does not exists */
} IfData;

IfData ifdata;

/* Prototypes */
void center(short int line, char *strg, ...);
void chcolor(int x, int y);
int update_stat_combined(void);
int update_stat_split(void);
int update_stat_large(void);
int update_stat(int);
void update_info(int);
void led_on(unsigned int who);
void led_off(unsigned int who);
int is_online(void);
int get_stat(void);
void clear_info(void);
void ip_address(int display);
void get_time (void);
void do_exit (int, char *, short int);
void usage (int, char **);
void draw_face (int, char *);
void zero_stats (void);
void slurm_shutdown (int);

/* Variables Declarations */
#define TYPE_MEGA 0
#define TYPE_GIGA 1
int data_type = TYPE_MEGA;
long refreshdelay = REFRESH_DEFAULT;

/* internal database status */
/* possible modes are: */
#define DB_STATUS_STARTUP 1
#define DB_STATUS_RUNNING 2
#define DB_STATUS_REINIT  3
int db_status = DB_STATUS_STARTUP;

#ifdef __linux
static FILE *proc_net_dev;
#endif
/* End of Variables Declarations */

/* max speed in graph */
#define GRAPHSINGLE_WIDTH 66
#define GRAPHSINGLE_HEIGHT 10

/* define graph height for split screen graphs */
#define GRAPHSPLIT_HEIGHT 7
#define GRAPHSPLIT_WIDTH 66

#define GRAPHCOMBINED_WIDTH 66
#define GRAPHCOMBINED_HEIGHT 14

/* new max height */
#define GRAPH_HEIGHT 12
#define GRAPH_WIDTH 66

/* large split mode graph height */
#define GRAPHLARGE_HEIGHT 10

/* rx is higher than need as we use it for the combined view */
int rx_graph[GRAPHSPLIT_WIDTH][GRAPHCOMBINED_HEIGHT];
float rx_speedarray[GRAPHSPLIT_WIDTH];
static float rx_maxspeed;
static float rx_minspeed;
static int rx_maxspeedpos;
static int rx_minspeedpos;
int rx_overallmax;

int tx_graph[GRAPHSPLIT_WIDTH][GRAPHCOMBINED_HEIGHT];
float tx_speedarray[GRAPHSPLIT_WIDTH];
static float tx_maxspeed;
static float tx_minspeed;
static int tx_maxspeedpos;
static int tx_minspeedpos;
int tx_overallmax;

/* combined view */
static float comb_maxspeed;
static float comb_minspeed;
int comb_overallmax;
int solcount;

int ledenabled;	/* TX/RX led enabled */


src/if_media.c
/******************************************************************************
 *
 * src/if_media.h - part of slurm
 *
 * this file handles basic network information functions for all
 * operating systems.
 *
 *****************************************************************************
 * $Id: if_media.c,v 1.35 2004/09/27 13:08:59 hscholz Exp $
 *****************************************************************************/

#if defined (__FreeBSD__) || defined (__OpenBSD__) || defined (__NetBSD__) || defined (__MicroBSD__) || defined (__APPLE__)
#define MEDIA_H_SUPPORTED
#endif

#if defined (__OpenBSD__) || defined (__NetBSD__) || defined (__MicroBSD__) || defined (__APPLE__)
#define NON_FreeBSD 1
#define WIRELESS 1
#else
#if defined(__FreeBSD__)
#if (__FreeBSD_version >= 450000)
#define WIRELESS 1
#endif
#else
#define WIRELESS 0
#endif
#endif

#if defined(__linux__)
#include <linux/sockios.h>
#include <linux/ethtool.h>
#endif

/******************************************************************************
 *
 * get_if_speed()
 *
 * determine current interface speed, needs interface name as argument
 * return the interface speed as an integer. unit: kbit/s
 * in case of error return ERR_IFACE_NO_SPEED
 *
 * tested/supported operating systems:
 *
 *  - FreeBSD
 *  - OpenBSD
 *  - NetBSD
 *  - MicroBSD (99% OpenBSD)
 *
 *****************************************************************************/

#ifdef MEDIA_H_SUPPORTED
int get_if_speed(char *ifstring)
{
    int speed = ERR_IFACE_NO_SPEED;
    int s;                      /* socket */
    struct ifmediareq ifmr;
    int *media_list;
    int type, physical;

    if ((s = socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
        fprintf(stderr, "cannot create socket!\n");
        return ERR_IFACE_NO_SPEED;
    }

    memset(&ifmr, 0, sizeof(ifmr));
    strncpy(ifmr.ifm_name, (char *) ifstring, sizeof(ifmr.ifm_name));

    if (ioctl(s, SIOCGIFMEDIA, (caddr_t) & ifmr) < 0) {
        fprintf(stderr,
                "interface does not support SIOCGIFMEDIA ioctl()\n");
        return ERR_IFACE_NO_SPEED;
    }

    if (ifmr.ifm_count == 0) {
        fprintf(stderr, "%s: no media types?\n", (char *) ifstring);
        return ERR_IFACE_NO_SPEED;
    }

    media_list = (int *) malloc(ifmr.ifm_count * sizeof(int));
    if (media_list == NULL)
        fprintf(stderr, "malloc() error in if_media.c\n");
    ifmr.ifm_ulist = media_list;

    if (ioctl(s, SIOCGIFMEDIA, (caddr_t) & ifmr) < 0) {
        fprintf(stderr, "ioctl(SIOCGIFMEDIA) failed\n");
        return -1;
    }

    /*
     * define type and physical
     *
     * bits:
     *      0-4 Media variant
     *      5-7 Media type
     *
     */

    type = ifmr.ifm_active & 0xf0;
    physical = ifmr.ifm_active & 0x0f;

#ifdef MEDIADEBUG
    printf("      all: %6d\n", ifmr.ifm_current);
    printf("   active: %6d\n", ifmr.ifm_active);
    printf("   status: %6d\n", ifmr.ifm_status);
    printf("     type: %6d\n", type);
    printf("     phys: %6d\n", physical);
    printf("if active: %6d\n", ifmr.ifm_active & IFM_ACTIVE);

#ifdef __APPLE__
    printf("    count: %6d\n", ifmr.ifm_count);
#endif                          /* Apple debugging */
#endif

    /* switch type */
    switch (type) {
        /* Ethernet */
    case IFM_ETHER:
        switch (physical) {
#ifdef __FreeBSD__
#if __FreeBSD__ <= 4
        case IFM_1000_FX:
        case IFM_1000_TX:
#endif
#endif
        case IFM_1000_SX:
        case IFM_1000_LX:
        case IFM_1000_CX:
#ifdef IFM_1000_T
        case IFM_1000_T:
#endif
            speed = 1000 * 1000;
            break;
        case IFM_100_TX:
        case IFM_100_FX:
        case IFM_100_T4:
        case IFM_100_VG:
        case IFM_100_T2:
            speed = 100 * 1000;
            break;
        case IFM_10_T:
        case IFM_10_2:
        case IFM_10_5:
        case IFM_10_FL:
        case IFM_10_STP:
            speed = 10 * 1000;
            break;
#if defined(__OpenBSD__) || defined (__MicroBSD__) || defined (__NetBSD__) || defined (__APPLE__)
        case IFM_HPNA_1:
#else
#if __FreeBSD__ <= 4
        case IFM_homePNA:
#endif
#endif
            speed = 1 * 1000;
            break;
        default:
            speed = ERR_IFACE_NO_SPEED;
            break;
        }                       /* end switch physical */
        break;
        /* FDDI interfaces */
        /* fpa doesn't seem to support SIOCGIFMEDIA on FreeBSD
         * so we won't get here but anyway ...
         */
    case IFM_FDDI:
        switch (physical) {
        case IFM_FDDI_SMF:
        case IFM_FDDI_MMF:
        case IFM_FDDI_UTP:
            speed = 100 * 1000;
            break;
        default:
            speed = ERR_IFACE_NO_SPEED;
        }
        break;
#if WIRELESS
        /* IEEE 802.11 wireless interfaces */
    case IFM_IEEE80211:
        switch (physical) {
        case IFM_IEEE80211_FH1:
        case IFM_IEEE80211_DS1:
            speed = 1 * 1000;
            break;
        case IFM_IEEE80211_FH2:
        case IFM_IEEE80211_DS2:
            speed = 2 * 1000;
            break;
        case IFM_IEEE80211_DS5:
            speed = (int) 5.5 *1000;
            break;
        case IFM_IEEE80211_DS11:
            speed = 11 * 1000;
            break;
#if WIRELESS && defined(__FreeBSD__) && (__FreeBSD_version >= 460102)
        case IFM_IEEE80211_DS22:
            speed = 22 * 1000;
            break;
#endif
#if WIRELESS && \
	(defined(__FreeBSD__) && (__FreeBSD_version >= 500111)) || \
	(defined(__NetBSD__) && (__NetBSD_Version_ > 106020000))
        case IFM_IEEE80211_OFDM6:
            speed = 6 * 1000;
            break;
        case IFM_IEEE80211_OFDM9:
            speed = 9 * 1000;
            break;
        case IFM_IEEE80211_OFDM12:
            speed = 12 * 1000;
            break;
        case IFM_IEEE80211_OFDM18:
            speed = 18 * 1000;
            break;
        case IFM_IEEE80211_OFDM24:
            speed = 24 * 1000;
            break;
        case IFM_IEEE80211_OFDM36:
            speed = 36 * 1000;
            break;
        case IFM_IEEE80211_OFDM48:
            speed = 48 * 1000;
            break;
        case IFM_IEEE80211_OFDM54:
            speed = 54 * 1000;
            break;
        case IFM_IEEE80211_OFDM72:
            speed = 72 * 1000;
            break;
#endif
        default:
            speed = ERR_IFACE_NO_SPEED;
            break;
        }
        break;
#endif
    default:
        speed = ERR_IFACE_NO_SPEED;
    }                           /* end switch type */

#ifdef MEDIADEBUG
    printf("    speed: %6d\n", speed);
#endif
    return speed;
}
#elif defined (__HPUX__)
int get_if_speed(char *ifstring)
{
    int speed = ERR_IFACE_NO_SPEED, buffer, fd, val, ret = -1;
    unsigned int len, i;
    struct nmparms params;
    mib_ifEntry *if_buf;

    for (i = 0; i <= ifdata.if_amount; i++) {
        if ((fd = open_mib("/dev/lan", O_RDWR, i, 0)) >= 0) {
            if ((if_buf =
                 (mib_ifEntry *) malloc(sizeof(mib_ifEntry))) != 0) {
                params.objid = ID_ifEntry;
                params.buffer = if_buf;
                len = sizeof(mib_ifEntry);
                params.len = &len;
                if_buf->ifIndex = i + 1;
                if ((ret = get_mib_info(fd, &params)) == 0) {
                    if (i + 1 == ifdata.if_id)
                        if (if_buf->ifOper == 1)
                            speed = if_buf->ifSpeed / 1000;
                        else
                            speed ERR_IFACE_DOWN;
                }
            }
        }
        free(if_buf);
        close_mib(fd);
    }
    return speed;
}
#elif defined (__Solaris__)
/******************************************************************************
 *
 * Solaris interface speed detection
 *
 *****************************************************************************/
int get_if_speed(char *ifstring)
{
    int speed = ERR_IFACE_NO_SPEED;
    kstat_t *ksp;
    kstat_named_t *knp;
    kstat_ctl_t *kc;

    if ((kc = kstat_open()) == NULL)
        return ERR_IFACE_NO_SPEED;

    ksp = kstat_lookup(kc, NULL, -1, ifstring);
    if (ksp && kstat_read(kc, ksp, NULL) >= 0) {
        knp = (kstat_named_t *) kstat_data_lookup(ksp, "ifspeed");
        if (knp)
            speed = (int) knp->value.ui64 / 1000;
    }
    kstat_close(kc);

    return speed;
}
#elif defined(__linux__)
int get_if_speed(char *ifstring)
{
    int speed = ERR_IFACE_NO_SPEED;
    int sock;
    struct ifreq ifr;
    struct ethtool_cmd edata;
    int rc;

    sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_IP);
    if (sock < 0) {
        perror("socket");
        fprintf(stderr, "Cannot create socket to get interface speed.");
        return -1;
    }

    strncpy(ifr.ifr_name, ifstring, sizeof(ifr.ifr_name));
    ifr.ifr_data = (void *)&edata;

    edata.cmd = ETHTOOL_GSET;

    rc = ioctl(sock, SIOCETHTOOL, &ifr);
    if (rc < 0) {
        perror("ioctl");
        fprintf(stderr, "Failed to get interface speed.");
        return -1;
    }
    if (ethtool_cmd_speed(&edata)) {
        speed = ethtool_cmd_speed(&edata) * 1000;
    }

    return speed;
}
#else
int get_if_speed(char *ifstring)
{
    ifstring++;                 /* ugly hack to prevent compiler warning on Linux */
    return ERR_IFACE_NO_SPEED;
}
#endif
